### PR TITLE
Fixed Lagrangian Particle boundary condition enforcement issue #2681

### PR DESCRIPTION
**Pull request by jagoosw**
Fixes Issue #2681

### CODE REVIEW

**Comment by glwagner** [comment]
Nice!

Is there a test we could add that might catch this bug?

**Comment by jagoosw** [comment]
Now I've looked harder I've realized that this issue is only for bounded domains so my original fix is incorrect for periodic domains since they are [xmin, xmax) not [xmin, xmax] like bounded domains. Fixed this now.

I've changed the position of the particles in the existing tests so that they should be affected by the change if that works? For the x I put it on the edge so it gets advected out (and will have been looped round even without the fix), y near the boundary but should not get looped round, and for the z I've put it near the boundary where previously it will have been moved but should not have been.

**Comment by simone-silvestri** [comment]
Nice! Good catch

File: src/LagrangianParticleTracking/update_particle_properties.jl [hunk 1]
```diff
@@ -22,12 +22,47 @@ along a `Periodic` dimension, put them on the other side.
     return x
 end
 
+# Fallback if we skip one cell
+@inline adjust_coord(x, args...) where N = x
+
+@inline adjust_coord(x, nodefunc, i, ::Val{0} , grid, rest) = x
+@inline adjust_coord(x, nodefunc, i, ::Val{-1}, grid, rest) = nodefunc(Face(), i+1, grid) - (x - nodefunc(Face(), i+1, grid)) * rest
+@inline adjust_coord(x, nodefunc, i, ::Val{1} , grid, rest) = nodefunc(Face(), i, grid)   + (nodefunc(Face(), i, grid)   - x) * rest
```
Comment (glwagner) [inline src/LagrangianParticleTracking/update_particle_properties.jl:30]: `ifelse` might be a better design here

**Comment by simone-silvestri** [comment]
@jagoosw do you mind if I commit to try to fix the tests?

**Comment by jagoosw** [comment]
> @jagoosw do you mind if I commit to try to fix the tests?

Not at all! They were working before but something must have changed. Thanks!

**Comment by simone-silvestri** [comment]
@jagoosw it seems like everything is ok! When the tests pass I ll merge, good job

**Review by simone-silvestri — APPROVED** [review]

**Comment by jagoosw** [comment]
@simone-silvestri great, thank you!

File: src/LagrangianParticleTracking/update_particle_properties.jl [hunk 2]
```diff
@@ -50,17 +50,19 @@ position based on the previous position (we bounce back a certain restitution fr
...
+            zₚ = adjust_coord(zₚ, znode, kₒ, kᵈ, grid, restitution)
         end
     end
 
     return xₚ, yₚ, zₚ
 end
 
-@inline function update_particle_position!(particles, p, restitution, grid::AbstractGrid{FT, TX, TY, TZ}, Δt, velocities) where {FT, TX, TY, TZ}
+@inline maxnode(::Bounded, nodes, N) = nodes[N+1]
+@inline maxnode(::Periodic, nodes, N) = nodes[N]
```
Comment (glwagner) [inline src/LagrangianParticleTracking/update_particle_properties.jl:63]: Missing @inbounds

**Pull request merged**